### ISSUE TITLE
Overarching vision for plot object and backend changes

### ISSUE THREAD

**Issue by hamogu** [sherpa/sherpa#1375]
The goal of this issue is to lay out, in narrative text, the broad line of planned plotting backend improvements; in that sense this text is a design document similar to the grouping/filtering word docs we've been looking at. At the completion of this project, essentially all text in this issue will have been transfered into the sherpa docs.

**I will edit this top description to adjust to any comments that are given in this issue. I take a "thumbs up" as general approval. Details can be discussed in individual issues, but leave comment here for big-picture suggestions.**

This issue is meant as a complement to the project tracking in https://github.com/sherpa/sherpa/projects/6. The project board tracks individual issues, but does not have the narrative text that I'm dumping into this issue for discussion. Individual issues are linked from the project board and therefore not all issues and PRs are listed here individually. 

## Step 1: Backends for plotting
We want backends to be cleanly separated from Sherpa code; they were always supposed to be but since we did have only one backend for a while and did not test for that, dependencies have snug in (#618, https://github.com/sherpa/sherpa/pull/1356#issuecomment-985764507). This will enable cleaner testing (#1229), hot-swapping backends (#348, #1191), simplify the implementation of any new backends (#499, #1356), and simplify maintenance of existing backends.

### Backends as objects, not modules
I proposal that backends should be instances of a class, not modules. There are multiple advantages:
- Objects can use inheritance - see how much code is duplicated in the pylab and the new bokeh backend in #1356 (as of 12/3/21 - #1356 will probably be rebased and re-worked to reduce that). 
- New backends can inherit from existing ones, which simplifies implementation as capabilities can be added piece by piece.
- Users can subclass a backend. For example, I would like to change the way error bars are displayed in matplotlib and show them as shaded regions instead of separate error bars. It's much more Pythonic to subclass, then to moneky-patch the exiting module. This is admittedly a power-user feature, but one that can simplify the use of sherpa as a library for others (like I intend for TGCat and others tried for gammapy in the past).
- Its easier to assign a new object then to switch out a module.
- Sherpa's plotting assumes a stateful backend with the notion of a "current plot" as in `plt.gca`. That is one approach many plotting packages provide (e.g. pylab in matplotlib) and that works well for Sherpa. However, for plotting packages that are more OO, the backend need to track the "current figure" etc. That could be done as a module-level variable in `sherpa.plot.XXX_backend`, but is much more natural as an instance attribute.

On the UI layer, the change from modules to objects can be invisible. Based on the setting in `.sherpa.rc`, an objects of e.g. `PylabBackend` can be initialized and do all the plotting without the users knowing how exactly it is implemented.

### Backend interface
**text is sketched out here as an idea for now and so that I don't loose the thought, but will be moved to a separate PR in the next few days to be discussed in detail there**

I propose that each backend shall support the following five plotting functions, where "support" means "has to provide these functions and accept a standard list of arguments without crashing or raising an exception". We explicitly allow for backends that implement some of these as a no-op, e.g. because the underlying plotting library does not support 2D data. In that case, the backend would typically issue a warning.

- plot (for scatter plots with `markerstyle` set, for line plots with `linestyle`set, and for errorbars with `xerr` or `yerr` set); accepts x,y data with optional xerr, yerr. Data can be scalar (for a single marker), or array-like. 
- histogram (similar to plot, but with "histogram-style" lines); accepts xlo, xhi, y data with optional xerr, yerr.
- vline/hline (for lines across the entire axes, used e.g. to mark the ratio=1 line)
- contour
- image

For each plotting function, there is a set of mandatory arguments that tweak the style of the plots. These arguments are backend-independent and each backend has to work with those. Treating them as a no-op is OK (e.g. if a plotting package does not allow for changes of the linestyle), but they cannot raise exception. 

[list to be spelled out in detail in docs, not here]
For example, `plot` has to accept a `linestyle` argument which can take at least the following values: ['solid', 'doted', 'dashed', '-', ':', '-.', None (for do not plot line)]. Depending on the backend, it *may* accept additional values. We define a minimum list of arguments so that we can define plots in the documentation and the tests that work for every backend. The list of backend-independent values that need to be accepted is based on chips and matplotlib heritage. That way, existing code that uses common options will work on every plotting backend. This matches the Sherpa 4.14 implementation that already accpets, e.g. `linestyle="solid"` in the pylab backend and translates that to `linestyle='-'` for matplotlib.

Each backend has additional functions that are called before, during and after interactive plots (`begin`, `onerror`, and `end`), and for the setup of multi-panel plots [those are taken essentially unchanged from the 4.14 version, so can copy from code into specs]

Backends need to have methods for
- `as_html_XXX` (where XXX is a plot type) that are used for interactive display in the notebook with `_repr_html_`. These functions take a plot object and return an html representation as a string.
- `get_XXX_plot/hist_prefs` (where XXX is a plot type) which return a dictionary of preferences that is used for displaying this plot.

Both these interfaces are not extensible for users and hard to maintain for developers since every new plot type requires adding a new set of functions for every backend. Inheritance helps a little, since modifiying the code for the base class is sufficient to get it working. Still, this interface needs to be improved, but that is done in step 2; it is not a lien on developing, merging, or releasing the code discussed so far.

Lastly, there is a simple helper method to latex formatting strings.

At this stage, sherpa will provide the following backends that work with the existing plotting objects:
- pylab (included in 4.14) - for matplotlib
- bokeh (drafted in #1356) - for bokeh
- dummy (included in 4.14) - no dependencies beyond sherpa to guarantee that plot objects can be used even if no plotting package is installed. Output can be ASCII or no-op.
- minimalbackend (no PR yet) - Similar to dummy, but raises an error if backend-specific options are used. For testing to ensure that tests and docs only use backend independent options.

## 2) Changes to the plot objects

[will edit text and fill in in the next few days. Too late today. Check back later.]

## 3) Plotting in the UI
There is a separate project tracking for plotting in the UI (https://github.com/sherpa/sherpa/projects/4). Many of the issues in that category deal with individual plots (e.g. add a specific plot type or change a specific option) and will also impact the OO ion some way (e.g. define a new plot class), but those a minor changes in principle that don't have to be discussed here - see the issues linked to the project board.

**Comment by wmclaugh** [comment]
@hamogu-
Thanks for adding this document. It is very useful and necessary for expanding/extending the plotting functionality. I agree with your plans for use of derived classes, heritage support for chips/matplotlib, and trying to avoid over engineering. I think It would be useful to have a follow up design meeting before you dive too far into implementing the bokeh backend.

Do you have class diagrams and descriptions of the relationships between proposed classes. Is there any distinction between a figure and objects such as plots (for example how is overplotting two different datasets handled).  What about bound axes for strip plots or associated plot (plot_fitdelchi). Do you anticipate having a baseline annotation class from which things like vline, hline, labels, regions, or other markers may belong?

**Comment by hamogu** [comment]
overplotting and linking axes is still missing from the document - see the note that says "[define interface here]". I'm not quite sure what to do, but I will continue to edit to make a suggestion. Essentially, our options are keeping it (very similar to) what we have now, or make a radical change. If we want to, now it the time to do so, but it's not clear to me if it's worth the effort.

Note that the bokeh backend in #1356 is fully functional, so it's too late for "before I dive too far into implementing" - but I would not be able to write this document, had I not implemented a backend already! There will probably be changes in #1356 based on the discussion here, but that's incremental.

**Comment by DougBurke** [comment]
I came across https://gaphor.org/ this morning which might help with setting up UML-like diagrams for the classes. However, having to learn enough UML to create this sort of thing is a big ask.

**Comment by hamogu** [comment]
People in astropy have occasionionally used https://draw.io which is pretty intuitive. However, I this particular case, I don't think we need a diagram. I'm not proposing an intricate class hierarchy, just a single class. Instead, I'm going to edit the text above to make that clear (and add a PR that converts `pylab_backend` into `class PylabBackend()`).

**Comment by hamogu** [comment]
@wmclaugh I've now written all I had in mind at this point. If you go back to the issue, you'll find the proposal expanded significantly. It explains why I don't think a diagram of classes is necessary here (if would probably help, but I think it's not worth the effort) and it fills in some of the other comments you had (mostly I suggest: Keep current API, so I did not write it down here because it's already documented in the existing backend classes).

**Comment by DougBurke** [comment]
Sorry - I've not read through the details recently, but I wanted to bring up

- #989
- #1021
- #1300 may fit into this (I'm not sure)

The first two are times where you would want to provide options to the data-preparation part of the plot objects rather than the plot part - that is the `prepare` method rather than the `plot` method - and the current system doesn't really support that.As mentioned I'm not 100% sure #1300 is the same (I am not sure what a general user would want, but I think that it again is in the preparation stage).

Does this fit in here, or can it be orthogonal?

**Comment by hamogu** [comment]
I've looked at all issues labelled "visualiaztion" when I wrote this and purposefully created a new project called "backend and plot objects" and I decided to leave those three issues out in and the "plot (UI)" project even though they are touching the plot objects and not just the session UI. They have nothing to do with the backend, it's all about sherpa data structures and sherpa functions. As I said above, most of the steps here are roughly orthogonal, but still will lead to conflicts when imported in parallel because (i) there are some small interactions, and (ii) they will touch the same files so, conflicts would come up in import statements, the text for the docs etc.  - that's why I suggest to implement it in steps in practice.

That said, I think the idea to have to sets of preferences, one for plotting and one for the prepare phase (https://github.com/sherpa/sherpa/issues/889#issuecomment-709579902) is promising and could be included either in step 3 or 4 of the suggested plan above. While saying that, I don't promise to implement all that's listed in https://github.com/sherpa/sherpa/projects/4, while I do promise to implement steps 0-3 above and, while doing so, to  consider all issues listed in https://github.com/sherpa/sherpa/projects/6. That won't close all of them, but a good fraction of it. 

I think another step could be

## Step X: Hierarchy of plots and composition of plots
which could contain issues like the above, it could mean defining running `plot_fit` as `plot_data + plot_model(overplot=True)` instead of having a separate objects for that etc. 

I have more ideas about that, but I purposefully left them out if this discussion, because I want this to be about "backends and how backend and plot objects interact." I'm interested to continue improving our visualization after we are done with this issue, but let's deal with the backends first.

**Comment by DougBurke** [comment]
The reason why I mentioned the issues above was that the current code relies on accessing the backend for the keywords they need, but this is only for the `plot` call, not `prepare`. I know you have thoughts on how the backend-specific options are set up/found/..., but I just wanted to mention this issue to think about when working on the backends.

**Comment by hamogu** [comment]
@wmclaugh I think I can answer the question about "an annotation class" now. Chips had an annotation class, and so does matplotlib [*], and bokeh. However, our sherpa backend code does not and, as far as I can tell, never had an annotation class. All it does have is a function, that says "generate this annotation in that backend". In the actual backend code (i.e. matplotlib or chips or boekh), that might be represented as an object or not, but sherpa itself does not know or care about that.

[I apologize to use the term "backend" for both the actual backend module (matplotlib) and also for the `sherpa.plot.pylab_backend` module. It's not a fortunate terminology, but it's what we are using in the code right now. Maybe we should rename it to `sherpa.plot.pylab_adapter`]

[*] I'm simplifiying. Matplotlib has the Artist class from which annotation derive, but it's not only annotations that derive from that.
